design/design-doc-composition-functions-extra-resources.md

```diff
@@ -0,0 +1,1196 @@
+# Composition Functions - Extra Resources
+
+* Owner: Philippe Scorsolini (@phisco)
+* Reviewers: TBD
+* Status: Draft
+
+## Background
+
+Composition Functions extend Crossplane to support a new way of configuring
+how to reconcile a Composite Resource (XR). Each Function is a gRPC server.
+Crossplane sends state to a Function via a `RunFunctionRequest` RPC. The
+Function is intended to return the desired state via a `RunFunctionResponse`
+RPC.
+
+Currently, a Composition author who needs to access information from a resource
+external to their Composition has the following options:
+- [**BETA**] Create an `ObserveOnly` resource from their Composition and
+  reading the required information from its `status`.
+- [**ALPHA**] Create an `EnvironmentConfig`, either manually or from another
+  Composition, with the required information and reference it from their
+  Composition via `spec.environment.environmentConfigs`.
+- Use MR-level referencing, if supported for the source and target of the
+  information, being implemented by the provider.
+
+Each of these strategies has its drawbacks:
+- The creation of `ObserveOnly` resources could result in a significant number
+  of resources being created.
+- The use of `EnvironmentConfigs` requires the Composition author to create and
+  maintain a separate resource, which can be cumbersome and error-prone.
+  However, it has to be said that there is value in the clean separation of concerns
+  between the `EnvironmentConfig`'s producer and consumer(s).
+- Using MR-level referencing is not always possible, as it requires explicit
+  provider support, and hardly ever crosses a single provider's (family) boundaries.
+
+As per the Composition Functions' [specification]:
+```text
+A Function MUST NOT assume it is deployed in any particular way, for example
+that it is running as a Kubernetes Pod in the same cluster as Crossplane.
+
+A Function MUST NOT assume it has network access. A Function SHOULD fail
+gracefully if it needs but does not have network access, for example by
+returning a Fatal result.
+```
+
+A clear consequence of the above assumptions is that a Function should not
+assume access to the Kubernetes API. Which means, that letting Functions fetch
+external resources on their own is not an option, Crossplane needs to provide
+them.
```

Comment:
I am not against the proposed design in general but is it a correct interpretation of the specification statement?

> A Function SHOULD fail
gracefully if it needs but does not have network access

sounds to me that we can expect the scenarios when Function requires network access but should fail gracefully in case it does not have it.

It does not sound to me that the network access should be fully avoided.

Reply:
Sure, that's why I say Functions SHOULD NOT fetch external resources on their own. They obviously can, but if they do, running them in less standard ways, e.g. `crossplane beta render`, would be much harder. So, we should at least provide some kind of native mechanism to allow doing that without breaking our assumptions for an "easy-to-run" Function.